### PR TITLE
feat: Added NOAA weather and BDIFF wildfire history datasets

### PR DESCRIPTION
**Pull request by frgfm**
This PR is based on #3 with several simplifications to ease the later development:
- added implementations of NOAA weather dataset and BDIFF history
- added a fallback mechanism for remote data sources
- added the `datasets` module documentation
- updated unittests accordingly (though they could still be expanded)
- added an example script, that you can run with : `python scripts/example_weather_wildfire.py`
- fixed license mentioned in setup.py

Here is how the documentation renders:
pyronear_ds_docs

The best course of action might be then to split up #3 into multiple PRs for the models and perhaps dataset merge (i.e. for now, I didn't create a real MergeDataset object, but it's built up in the example script)
Any feedback is welcome!

### CODE REVIEW

**Comment by dataJSA** [comment]
It works for me thanks @frgfm

**Comment by chloeskt** [comment]
Hello @frgfm, thank you for the PR. 
I am currently reviewed it, but it seems that our classes do not work correctly. I just ran the script example `example_weather_wildfire.py` and it seems to me that the merger using `df = pd.merge(weather, history, left_on=['DATE', 'code'], right_on=['date', 'Département'], how='left')` does not work. We shouldn't be getting NaN values, should we ? 

Capture d’écran 2020-11-04 à 09 19 32
Capture d’écran 2020-11-04 à 09 10 20

**Comment by dataJSA** [comment]
Concerning the dataset merging should we abstract it away from the user. What I mean by that is that we could keep all the function needed to fetched and merge the data under the `utils.py` module host the clean dataset with the release and only have one dataset class. 

Let me know what you think !

**Comment by chloeskt** [comment]
> Concerning the dataset merging should we abstract it away from the user. What I mean by that is that we could keep all the function needed to fetched and merge the data under the `utils.py` module host the clean dataset with the release and only have one dataset class.
> 
> Let me know what you think !

I'm not sure, I think that the user should have the choice on how to merge the dataset. One might want to merge according to the departement geometry, and another might want to merge according to the nearest weather station such as Camille suggested in #3

**Comment by frgfm** [comment]
@chloeskt @dataJSA About the merging & NAs, I think we might have to rethink the date range selection. Here is the problem : 
1. in #3, the selection is made taking the largest date range of both data sources
2. here, I take the largest intersecting date range
3. in both cases, while the weather is always reported for a given location and date, if there is no wildfire in this location & date, there won't be any entry in the data. Meaning that we should consider the wildfire history date range as the date range of the extract request and not the actual data.

The problem with 1 is that we can merge data that are not at all temporarily compatible, and the issue here is that we might get a smaller date range than expected.

My suggestion would be: I remove for now the extra columns that do not bring any extra information, and we deal with date range selection in a separate issue/PR (since this might require to be given much thoughts). What do you think?

**Comment by chloeskt** [comment]
> @chloeskt @dataJSA About the merging & NAs, I think we might have to rethink the date range selection. Here is the problem :
> 
>     1. in #3, the selection is made taking the largest date range of both data sources
> 
>     2. here, I take the largest intersecting date range
> 
>     3. in both cases, while the weather is always reported for a given location and date, if there is no wildfire in this location & date, there won't be any entry in the data. Meaning that we should consider the wildfire history date range as the date range of the extract request and not the request response data. (which will be handled when we move to API requests for the different data sources)
> 
> 
> The problem with 1 is that we can merge data that are not at all temporarily compatible, and the issue here is that we might get a smaller date range than expected.
> 
> My suggestion would be: I remove for now the extra columns that do not bring any extra information, and we deal with date range selection in a separate issue/PR (since this might require to be given much thoughts). What do you think?

I am not sure to get your point. In #3 the selection is made by taking the smallest date range. No matter which dataset has the smallest date range, the wildfire history dataset will contain less rows (as in less dates). However by doing a "left" merge with the weather dataset as the left dataset, we can keep all dates in between the smallest range and only have Statut = 1 for the dates on which there is a fire (i.e. we have an observation in the wildfire dataset), and 0 otherwise (for the dates for which we have historic weather conditions).
Therefore, we shouldn't get a smaller date range than expected. Basically I did this in #3 because I have more recent data in the weather dataset. Also I think it can be useful to have geographic data points on which no fire occured, which we get by using all the weather dataset and doing a left merge (if this dataset in the left one). 
I don't know if I'm clear ... 

But yes, I agree with you, we can deal with date range selection in a separate PR 👍

**Comment by frgfm** [comment]
@chloeskt I might be missing something because to me, having the smallest date range does not mean having a range that overlaps both datasets. And that's where I'm not following :thinking: 
But you know the data far better than I do, so if you have an edit suggestion on https://github.com/pyronear/pyronear-datascience/blob/dataset-draft/pyronear_ds/datasets/utils.py, I'm all ears!

**Comment by chloeskt** [comment]
@frgfm oh okay, I understand what you meant now. My bad. I wasn't looking at taking the intersection per se; sorry. 
However there is still an issue. When I print the intersection of both datasets, it goes from 01-01-2018 to 30-12-2019. When I run the line `history = history[history.date.isin(date_range)]` , I get a very small `history` dataframe, it shouldn't be the case. The whole fire dataset starts on 2018-01-08. 

I have yet to located the issue, or maybe I'm just going crazy 😅

Capture d’écran 2020-11-04 à 22 47 42

**Comment by chloeskt** [comment]
Sorry, I might have a bug on my computer, but each time I rerun the `example_weather_wildfire.py` script I get different results ... I did not change a thing and I got a different result as you can see below: it is normal ? I am the only one to get different results at each run ? 


Capture d’écran 2020-11-04 à 22 58 42

**Comment by frgfm** [comment]
@chloeskt strange, I got the same issue over here actually!
Weirdest thing is: I consistently have the same length of weather data, history data and the same intersecting date range, but after indexing, it does not yield the same thing :man_shrugging:

**Review by chloeskt — CHANGES_REQUESTED** [review]
Hi @frgfm I think I found the issue, it was with datetime object conversion. We need both `DATE` in the weather dataset and `date` in the wildfires one to be of the type `datetime64[ns]` in order to correctly to the merge. It wasn't the case before due to the addition of `.dt.date` which changes the dtype to `object`. We do not need to add `.dt.date` since datetime object have both date and time attributes and it does not impact the merge: when the format is `%Y-%m-%d` the time is always set to 00:00:00 so dates accross datasets are comparable. 

With the suggested modifications, you should get a merged dataframe with shape (97678, 29), a `Statut` value counts of 91471 zeros and 6207 ones. Finally, the percentage of missing values should be 0 everywhere except for `Département` and `date` where it should be `93.645447` (because we are doing a left merge on the weather dataset).

This yields the same results as my pipeline in #3

**Comment by chloeskt** [comment]
However I've noticed something for which I do not have any explanation yet: the number of declared wildfires before and after the merge is different was you can see below. We get roughly 2000 more wildfires after the merge, I wonder why. 
It's the same in #3 , but I think that it shouldn't be the case; what do you think ? 

Capture d’écran 2020-11-06 à 07 29 02

File: pyronear_ds/datasets/wildfires.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    def __init__(self, source_path: Optional[str] = None, use_cols: Optional[List[str]] = None) -> None:
+        if not isinstance(source_path, str):
+            # Download in cache
+            logging.warning(f"No data source specified for {self.__class__.__name__}, trying fallback.")
+            source_path = cfg.FR_FIRES_FALLBACK
+        if not isinstance(use_cols, list):
+            use_cols = self.kept_cols
+        data = pd.read_csv(source_path, sep=";", usecols=use_cols)
+        tmp = pd.to_datetime(data['Date de première alerte'], format="%Y-%m-%d %H:%M:%S", errors='coerce')
```
[the remark this answers is not in the document]
Comment (frgfm) [reply pyronear_ds/datasets/wildfires.py:29]: Actually when forcing `format="%Y-%m-%d"` on a date + time information, the time is kept so this was still generating errors. I used `.dt.normalize()` to set the time part to `00:00:00` for each entry, which seems to be fixing the problem entirely !

**Comment by frgfm** [comment]
Thanks for the investigation @chloeskt
I fixed the datetime normalization, and I'm now getting a consistent size of output dataset with the example script :ok_hand:

**Review by chloeskt — APPROVED** [review]

**Pull request merged**